oeps/oep-0028-content-theming-in-react.rst

```diff
@@ -0,0 +1,235 @@
+            </div>;
+        }
+    }
+
+    // Default Theme:
+    export const Header = _Header;
+    export const MainNav = _MainNav;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+Now if we want to customize our _Header component, we can easily do it like
+
+.. code-block:: js
+
+    // SiteLogo being updated in Header
+    class MyThemedHeader extends _Header {
+        public render() {
+            return (
+                <header>
+                    {/* Replace <SiteLogo/> with a fancy widget */}
+                    <MyCustomAnimatedLogoWidget/>
+                    <MainNav/>
+                    <UserAvatar/>
+                </header>
+            );
+        }
+    }
+    // Custom theme:
+    export const Header = MyThemedHeader;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+* We will provide support to add data or other components by using `props.children`. The `props.children` property is provided from React and can be extended to include specialized children like left and right for a Pane component. An example of this can be
+
+.. code-block:: js
+
+    const Header = (props) => {
+        return (
+            <h1>{props.title}</h1>
+            {props.children}
+        )
+    }
+
+    // use
+    <Header title={"Open Edx"} >
+        <h4>Open Source MOOC platform </h4>
+    </Header>
+
+Another example where specialized children are passed
+
+.. code-block:: js
+
+    const Pane = (props) => {
+        return (
+            <div className="pane">
+                <div className="pane-left">
+                    {props.left}
+                </div>
+                <div className="pane-right">
+                    {props.right}
+                </div>
+            </div>
+
+        )
+    }
+
+    // use
+    <Pane left={<Image />} right={<Explanation />} />
+
+
+* We will use function and placeholder to add additional content in customizable components. We can take an example of the above DefaultTheme and see _MainNav where it has support to add additional nav links by overriding `extraNavLinks` function.
+
+.. code-block:: js
+
+    // Customizable Main Navigation Area
+    class MyThemedNav extends _MainNav {
+        get extraNavLinks() {
+            return [
+                <a href="/about">About Us</a>,
+            ];
+        }
+    }
+
+    // Custom theme:
+    export const MainNav = MyThemedNav;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+* We will provide support for a global redux store, which will act as a central place to store data. This store will access to all data available to the system.
+
+* We will consider the layout of the data in the global store as an API. We will provide support to pre-fill the store with some common data like current user, current course, list of courses enrolled, etc. and add the less commonly needed data from REST API's using built-in redux actions. We will announce breaking changes if the layout of the data changes in store.
```

Comment:
I feel like the Open edX global redux store should be considered a stable API, but if themes want to read/write data that's not already part of that store, they should store it in their own separate redux store, and do their own fetching of data from REST APIs etc., so as not to conflict with the Open edX redux store, other themes, other plugins, etc. Thoughts?

Reply:
I don't think perhaps we can just namespace the stable vs unstable events and state. I also feel that instead of making the state itself stable API, the stable bit can be small selector methods that dig through the state to get the data you want. So you would have a function like:

```getUserProfileInfo = (state, userid) => state.users[userid].profile```

If the state shape changes, the worst case would be that this function becomes more complex and now pulls in data from multiple sources but still provides the same API.

Reply:
@xitij2000 @bradenmacdonald : From the discussion, there can be two approaches
* State as a stable API 
* OR State consisting of selector method as a stable API

I think we can keep Braden's point of considering the state as a stable API and any theme can create their own separate redux store if they are requesting some data which is not present in the Open edx redux store. In the [gist](https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115) Matjaz also [mentioned](https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115#gistcomment-2624625) about keeping the state as API.
Let me know if I am missing anything here?